### PR TITLE
Add initial support for wxBitmapBundle to wxRibbonButtonBar

### PR DESCRIPTION
**Pull request by Runtemund**
The first part of adding `wxBitmapBundle` support to `wxRibbonBar` in general.

This adds initial support for `wxBitmapBundle` to `wxRibbonButtonBar`.

The code was tested under W10, macOS 12.0.1 and openSuse 15.3.

I had to add a compiler switch because the behavior of `wxImageList` is not exactly the same on all platforms. Under UNIX it works with the logical size of a bitmap, but under MSW it seems, that it only works with the physical size. I figured out, if scaling factor of the bitmap is not 1.0, the ImageList cuts the right and bottom edge of the bitmap under W10. Probably this is an issue related to `wxImageList`.

### CODE REVIEW

File: include/wx/ribbon/art.h [hunk 1]
```diff
@@ -20,6 +20,7 @@
 #include "wx/font.h"
 #include "wx/pen.h"
 #include "wx/bitmap.h"
+#include "wx/bmpbndl.h"
```
Comment (vadz) [inline include/wx/ribbon/art.h:23]: Very minor, but
```suggestion
#include "wx/bmpbndl.h"
```
(we don't need both of them as the latter includes the former anyhow)

File: include/wx/ribbon/buttonbar.h [hunk 2]
```diff
@@ -202,6 +203,7 @@ class WXDLLIMPEXP_RIBBON wxRibbonButtonBar : public wxRibbonControl
     wxRibbonButtonBarButtonInstance* m_active_button;
 
     wxPoint m_layout_offset;
+    // Sizes are the logical size
```
Comment (vadz) [inline include/wx/ribbon/buttonbar.h:206]: I don't think we should this comment, _all_ sizes are in logical units in wx API, unless explicitly mentioned otherwise, so having it just here, and not in a million of other places just like this, is inconsistent and even somewhat misleading, as it implies this one is somehow different.
```suggestion
```

File: include/wx/ribbon/buttonbar.h [hunk 3]
```diff
@@ -156,6 +156,7 @@ class WXDLLIMPEXP_RIBBON wxRibbonButtonBar : public wxRibbonControl
                 wxRibbonButtonBarButtonState min_size_class);
     virtual void SetButtonMaxSizeClass(int button_id,
                 wxRibbonButtonBarButtonState max_size_class);
+    virtual void SetButtonBitmapSize(const wxSize& large_btn, const wxSize& small_btn);
```
Comment (vadz) [inline include/wx/ribbon/buttonbar.h:159]: IMHO would be more consistent to use `size`, rather than `btn`, in the parameter names here:
```suggestion
    virtual void SetButtonBitmapSize(const wxSize& size_large, const wxSize& size_small);
```
(of course, in the rest of wx API we use `camelCase` rather than `snake_case`, so it's already not very consistent, but like this it's at least locally consistent)

File: samples/ribbon/ribbondemo.cpp [hunk 4]
```diff
@@ -368,8 +368,8 @@ MyFrame::MyFrame()
         sizer_panelcombo2->SetMinSize(wxSize(150, -1));
 
         wxRibbonButtonBar* bar = new wxRibbonButtonBar(sizer_panel, wxID_ANY);
-        bar->AddButton(ID_BUTTON_XX, "xx", ribbon_xpm);
-        bar->AddButton(ID_BUTTON_XY, "xy", ribbon_xpm);
+        bar->AddButton(ID_BUTTON_XX, "xx", wxBitmap(ribbon_xpm));
```
Comment (vadz) [inline samples/ribbon/ribbondemo.cpp:371]: This doesn't need to be done in this PR, of course, but it would be nice if we could have a high resolution version of this (and all the other) XPMs or, ideally, an SVG.

File: src/ribbon/buttonbar.cpp [hunk 5]
```diff
@@ -1532,6 +1560,15 @@ wxRect wxRibbonButtonBar::GetItemRect(int button_id)const
     return wxRect();
 }
 
+void wxRibbonButtonBar::SetButtonBitmapSize(const wxSize& large_btn, const wxSize& small_btn)
+{
+    if (m_buttons.IsEmpty())
```
Comment (vadz) [inline src/ribbon/buttonbar.cpp:1565]: This should be an assert, i.e.
```suggestion
    wxCHECK_RET( m_buttons.IsEmpty(), "Must be called before adding any buttons" );
```
as silently ignoring the provided values is really not very developer-friendly.

File: src/ribbon/buttonbar.cpp [hunk 6]
```diff
@@ -978,8 +1004,10 @@ void wxRibbonButtonBar::CommonInit(long WXUNUSED(style))
     // Create() later.
     m_ribbonBar = GetAncestorRibbonBar();
 
-    m_bitmap_size_large = wxSize(32, 32);
-    m_bitmap_size_small = wxSize(16, 16);
+    // Note: size definition before 3.1.7 was always overriden by first added button
+    // in wxRibbonButtonBar::InsertButton()
+    m_bitmap_size_large = wxSize(0, 0);
+    m_bitmap_size_small = wxSize(0, 0);
```
Comment (vadz) [inline src/ribbon/buttonbar.cpp:1010]: Do we actually need to do this at all? They're (0,0) by default. If do initialize them, it would arguably make more sense to initialize them to `wxDefaultSize`.

I also don't think we need such historical comments as the one above, we have the VCS history for this.

File: src/ribbon/buttonbar.cpp [hunk 7]
```diff
@@ -406,29 +428,32 @@ wxRibbonButtonBarButtonBase* wxRibbonButtonBar::InsertButton(
...
     if(m_buttons.IsEmpty())
     {
+        bool setLarge = m_bitmap_size_large.x == 0 || m_bitmap_size_large.y == 0;
+        bool setSmall = m_bitmap_size_small.x == 0 || m_bitmap_size_small.y == 0;
+
         if(bitmap.IsOk())
         {
-            m_bitmap_size_large = bitmap.GetLogicalSize();
-            if(!bitmap_small.IsOk())
+            if(setLarge)m_bitmap_size_large = bitmap.GetPreferredLogicalSizeFor(m_ribbonBar);
```
Comment (vadz) [inline src/ribbon/buttonbar.cpp:446]:
```suggestion
            if ( setLarge )
                m_bitmap_size_large = bitmap.GetPreferredLogicalSizeFor(m_ribbonBar);
```
Please don't put compound statements all on the same line.

File: src/ribbon/buttonbar.cpp [hunk 8]
```diff
@@ -406,29 +428,32 @@ wxRibbonButtonBarButtonBase* wxRibbonButtonBar::InsertButton(
...
+                const wxBitmapBundle& bitmap_small,
+                const wxBitmapBundle& bitmap_disabled,
+                const wxBitmapBundle& bitmap_small_disabled,
                 wxRibbonButtonKind kind,
                 const wxString& help_string)
 {
     wxASSERT(bitmap.IsOk() || bitmap_small.IsOk());
     if(m_buttons.IsEmpty())
     {
+        bool setLarge = m_bitmap_size_large.x == 0 || m_bitmap_size_large.y == 0;
```
Comment (vadz) [inline src/ribbon/buttonbar.cpp:441]: Note: if we use `wxDefaultSize` initially, this could be written as `IsFullySpecified()`.

File: src/ribbon/buttonbar.cpp [hunk 9]
```diff
@@ -92,30 +81,39 @@ class wxRibbonButtonBarButtonBase
...
+                    const wxBitmapBundle &bundle_large,
+                    const wxBitmapBundle &bundle_large_disabled,
+                    const wxBitmapBundle &bundle_small,
+                    const wxBitmapBundle &bundle_small_disabled)
     {
+        double scale = ribbon->GetDPIScaleFactor();
+        wxSize szLarge = bitmap_size_large * scale;
+        wxSize szSmall = bitmap_size_small * scale;
+
+        wxBitmap bitmap_large = (bundle_large.IsOk()) ?
```
Comment (vadz) [inline src/ribbon/buttonbar.cpp:97]: We don't need `IsOk()` tests here, calling `GetBitmap()` on a null bundle returns a null bitmap anyhow and omitting these checks would make the code simpler and shorter.

File: src/ribbon/buttonbar.cpp [hunk 10]
```diff
@@ -130,15 +128,32 @@ class wxRibbonButtonBarButtonBase
 
         if ( bitmap_large.IsOk() )
         {
+            bitmap_large.SetScaleFactor(scale);
+            bitmap_large_disabled.SetScaleFactor(scale);
+
+            //Behavior of ImageList is not the same on different platforms
```
Comment (vadz) [inline src/ribbon/buttonbar.cpp:134]: We really need to isolate and fix this problem instead of working around it here.

File: include/wx/ribbon/buttonbar.h [hunk 11]
```diff
@@ -58,68 +58,68 @@ class WXDLLIMPEXP_RIBBON wxRibbonButtonBar : public wxRibbonControl
...
                 int button_id,
                 const wxString& label,
-                const wxBitmap& bitmap,
-                const wxBitmap& bitmap_small = wxNullBitmap,
-                const wxBitmap& bitmap_disabled = wxNullBitmap,
-                const wxBitmap& bitmap_small_disabled = wxNullBitmap,
+                const wxBitmapBundle& bitmap,
+                const wxBitmapBundle& bitmap_small = wxNullBitmap,
+                const wxBitmapBundle& bitmap_disabled = wxNullBitmap,
+                const wxBitmapBundle& bitmap_small_disabled = wxNullBitmap,
```
Comment (vadz) [inline include/wx/ribbon/buttonbar.h:82]: Sorry, this doesn't make sense, the bundle should contain both the normal and the small versions of the image. I.e. we need to keep the original function for compatibility, but the new one should take only `bundle` and `bundle_disabled` and not 4 bundles.

File: include/wx/ribbon/buttonbar.h [hunk 12]
```diff
@@ -58,68 +58,68 @@ class WXDLLIMPEXP_RIBBON wxRibbonButtonBar : public wxRibbonControl
...
```
Comment (vadz) [inline include/wx/ribbon/buttonbar.h:122]: Same comment as above (and also same comment for `SetButtonIcon()` below).

**Review by vadz — CHANGES_REQUESTED** [review]
Thanks for your work on this!

This is definitely a step in the right direction, but there are 2 problems:

1. We shouldn't use different bundles for the large and small images, they should be in the same one.
2. We need to fix the problem in `wxImageList` itself. If you could please explain what exactly it is, i.e. how can it be reproduced, it would be great. TIA!

File: include/wx/ribbon/buttonbar.h (hunk 3 above)
Comment (Runtemund) [reply include/wx/ribbon/buttonbar.h:159]: I am familiar with camel case, I have no problems to change it all by time.

File: samples/ribbon/ribbondemo.cpp (hunk 4 above)
Comment (Runtemund) [reply samples/ribbon/ribbondemo.cpp:371]: I just updated the example to make it work. There are also some other parts of wxRibbonBar where such updates are necessary, for example for drop down arrows... I would rather do it later.

File: src/ribbon/buttonbar.cpp (hunk 6 above)
Comment (Runtemund) [reply src/ribbon/buttonbar.cpp:1010]: I can delete it. It was just added for clarifying, why I removed the old "default" values. Initializing with 0,0 (by default) is in my opinion the best way, because it doesn't change actual default behavior and additional with `SetBitmapSize()` the developer can decide if we wants to use the default size or another size.

File: src/ribbon/buttonbar.cpp (hunk 9 above)
Comment (Runtemund) [reply src/ribbon/buttonbar.cpp:97]: OK, I will rethink this.

File: src/ribbon/buttonbar.cpp (hunk 8 above)
Comment (Runtemund) [reply src/ribbon/buttonbar.cpp:441]: OK, I will check this.

**Comment by Runtemund** [comment]
@vadz I tried to figure out, what is wrong with `wxImageList` and how to reproduce it. My simple idea was to write an unit test. I found an existing one which already is doing this job, but by default is not executed on my MSW machine. If I remove the switch `#ifdef wxHAS_DPI_INDEPENDENT_PIXELS` in `tests/graphics/imagelist.cpp` and the test is executed under W10 it fails here. On Linux not. Does this help?

**Comment by vadz** [comment]
The test fails because it explicitly checks that for bitmap with the scale factor 2 their logical size is twice smaller than their physical size, but it's not a problem per se, it's just how the API works (it's inconsistent between platforms, but we couldn't avoid having some inconsistency somewhere while still keeping compatibility, so this is how it is).

But it is possible to write the application code without relying on this, i.e. using either physical or logical size, as needed. It shouldn't be necessary to use `#ifdef`s in the code using `wxImageList` and it's still not clear to me why do you need to do it. Perhaps there is actually nothing to be fixed at `wxImageList` side, but this code should be changed.

**Comment by Runtemund** [comment]
Ok, if it is necessary to keep this inconsistency for compatibility reasons, I have to deal with it. I will switch to physical size internally instead of logical size. This should work...

File: include/wx/ribbon/buttonbar.h (hunk 12 above)
Comment (Runtemund) [reply include/wx/ribbon/buttonbar.h]: Done.

File: include/wx/ribbon/buttonbar.h (hunk 11 above)
Comment (Runtemund) [reply include/wx/ribbon/buttonbar.h]: Done.

File: src/ribbon/buttonbar.cpp (hunk 10 above)
Comment (Runtemund) [reply src/ribbon/buttonbar.cpp]: Done.

File: src/ribbon/buttonbar.cpp (hunk 8 above)
Comment (Runtemund) [reply src/ribbon/buttonbar.cpp]: Done.

**Comment by Runtemund** [comment]
From my point of view, open topics are done. Please review.

File: src/ribbon/buttonbar.cpp [hunk 13]
```diff
@@ -1532,6 +1569,15 @@ wxRect wxRibbonButtonBar::GetItemRect(int button_id)const
     return wxRect();
 }
 
+void wxRibbonButtonBar::SetButtonBitmapSize(const wxSize& large_btn, const wxSize& small_btn)
+{
+    wxCHECK_RET( m_buttons.IsEmpty(), "Must be called before adding any buttons" );
+
+    m_bitmap_size_large = large_btn;
+    m_bitmap_size_small = small_btn;
+    m_bitmap_size_small = small_btn;
```
Comment (vadz) [inline src/ribbon/buttonbar.cpp:1578]:
```suggestion
```

File: src/ribbon/buttonbar.cpp [hunk 14]
```diff
@@ -130,15 +112,25 @@ class wxRibbonButtonBarButtonBase
 
         if ( bitmap_large.IsOk() )
         {
+            // wxImageList handles logical size inconsistent between some platforms,
+            // which has to be kept for compatibility reasons.
+            // We use physical size for storing and ensure, that logical size is the same.
+            bitmap_large.SetScaleFactor(1.0);
```
Comment (vadz) [inline src/ribbon/buttonbar.cpp:118]: Sorry but does this really work under GTK/macOS, i.e. platforms where the scale factor is actually used when drawing the bitmap? If you have already tested it there and it does work, I must be missing something, but if not, I'm afraid this isn't the right thing to do.

**Review by vadz — COMMENTED** [review]
Thanks for the updates!

Formally the code looks fine now, but I really wonder how can this work under e.g. macOS because AFAIK we need to use the correct scale factor there when actually rendering bitmaps.

Could you please try to explain the actual problem you're facing with `wxImageList` once again? I just don't understand why can't we use it without doing all these scale manipulations. I.e.

1. When adding the first bitmap, determine the size of the image list to create.
2. When adding subsequent bitmaps, ask the bundle for a bitmap of the required size.

Why doesn't this work?
